worker/uniter/secrets/secrets.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2022 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package secrets
+
+import (
+	"sync"
+
+	"github.com/juju/charm/v9/hooks"
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+
+	coresecrets "github.com/juju/juju/core/secrets"
+	"github.com/juju/juju/worker/uniter/hook"
+	"github.com/juju/juju/worker/uniter/operation"
+	"github.com/juju/juju/worker/uniter/remotestate"
+)
+
+// SecretsClient is used by the secrets tracker to access the Juju model.
+type SecretsClient interface {
+	remotestate.SecretsClient
+	SecretMetadata(filter coresecrets.Filter) ([]coresecrets.SecretOwnerMetadata, error)
+}
+
+// Secrets generates storage hooks in response to changes to
+// storage Secrets, and provides access to information about
+// storage Secrets to hooks.
+type Secrets struct {
+	client  SecretsClient
+	unitTag names.UnitTag
+	stateRW UnitStateReadWriter
+	logger  Logger
+
+	secretsState *State
+	stateOps     *stateOps
+
+	mu sync.Mutex
+}
+
+// NewSecrets returns a new secrets tracker.
+func NewSecrets(
+	client SecretsClient,
+	tag names.UnitTag,
+	rw UnitStateReadWriter,
+	logger Logger,
+) (SecretStateTracker, error) {
+	a := &Secrets{
+		client:   client,
+		unitTag:  tag,
+		logger:   logger,
+		stateOps: NewStateOps(rw),
+	}
+	if err := a.init(); err != nil {
```

Comment:
This could just be written  as.

```
err := a.init()
return a, err
```

Reply:
nah, cause then we could not be returning the nil value on error, which is the idiomatic convention wherever possible
the "a" is misnamed, should be "s"